### PR TITLE
Vote refactor

### PR DESCRIPTION
**Pull request by Lumipharon**
## About The Pull Request
Refactors vote code to remove snowflakey spaghetti regarding gamemodes with specific maps.

Gamemodes now have whitelists and blacklists for ship and ground maps, making it easier in the future as new gamemodes and map dependencies are added.

Now instead of having specific bits of code to load a specific map, as an example voting combat patrol will trigger a ship map vote which will default to combat patrol base as it will be the only valid option.

## Why It's Good For The Game
Less sphaghetti

## Changelog
:cl:
refactor: Refactor vote code to be less spaghetti
/:cl:

### CODE REVIEW

**Comment by Lumipharon** [comment]
@TiviPlus This will require a config change. Combat Patrol Base (and I guess twin pillars) can now be votable like the rest of the maps, as they'll be blacklisted out of the votes when other modes are in play.

**Comment by TiviPlus** [comment]
Noted

File: code/datums/gamemodes/game_mode.dm [hunk 1]
```diff
@@ -44,6 +44,15 @@
...
 	var/silo_scaling = 1
 
+	///If the gamemode has a whitelist of valid ship maps
+	var/list/white_list_ship_maps
+	///If the gamemode has a blacklist of disallowed ship maps
+	var/list/black_list_ship_maps = list(MAP_COMBAT_PATROL_BASE, MAP_TWIN_PILLARS)
+	///If the gamemode has a whitelist of valid ground maps
+	var/list/white_list_ground_maps
+	///If the gamemode has a blacklist of disallowed ground maps
+	var/list/black_list_ground_maps =list(MAP_WHISKEY_OUTPOST)
```
Comment (TiviPlus) [inline code/datums/gamemodes/game_mode.dm:54]:
```suggestion
	var/list/black_list_ground_maps = list(MAP_WHISKEY_OUTPOST)
```

**Review by TiviPlus — CHANGES_REQUESTED** [review]
What exactly do I need to edit in config btw

File: code/controllers/subsystem/vote.dm [hunk 2]
```diff
@@ -241,10 +244,8 @@ SUBSYSTEM_DEF(vote)
 					var/players = length(GLOB.clients)
 					if(mode.time_between_round && (world.realtime - SSpersistence.last_modes_round_date[mode.name]) < mode.time_between_round)
 						continue
-					if(istype(mode, /datum/game_mode/civil_war) && SSticker.current_state < GAME_STATE_PLAYING && SSmapping.configs[SHIP_MAP].map_name != MAP_TWIN_PILLARS)
-						continue
-					if(istype(mode, /datum/game_mode/combat_patrol) && SSticker.current_state < GAME_STATE_PLAYING && SSmapping.configs[SHIP_MAP].map_name != MAP_COMBAT_PATROL_BASE)
-						continue
+					//if(mode.white_list_ship_maps && !(SSmapping.configs[SHIP_MAP].map_name in mode.white_list_ship_maps)) //if a map has a whitelist ship map that isn't currently selected, cannot vote to avoid more restarts
+					//	continue
```
[the remark this answers is not in the document]
Comment (Lumipharon) [reply code/controllers/subsystem/vote.dm:248]: Whoops forgot to delete.

**Comment by Lumipharon** [comment]
> What exactly do I need to edit in config btw

In the shipmap config, I believe that patrol base and twin pillars have "voteweight 0" which needs to be removed.

**Review by BraveMole — CHANGES_REQUESTED** [review]
It is whitelist and blacklist, one word.
Rest lgtm

File: code/controllers/subsystem/vote.dm [hunk 3]
```diff
@@ -136,30 +136,33 @@ SUBSYSTEM_DEF(vote)
...
-				SSmapping.changemap(VM, SHIP_MAP)
 			SSticker.save_mode(.) //changes the next game mode
 			if(GLOB.master_mode == .)
 				return
 			if(SSticker.HasRoundStarted())
 				restart = TRUE
 			else
-				var/datum/game_mode/current_gamemode = config.pick_mode(GLOB.master_mode)
+				var/ship_change_required
+				var/ground_change_required
```
Comment (BraveMole) [inline code/controllers/subsystem/vote.dm:146]: You should set them to false for clarity

**Review by BraveMole — APPROVED** [review]

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**